### PR TITLE
Bot API 5.0 

### PR DESCRIPTION
**Pull request by Fedorus**

### CODE REVIEW

File: src/Telegram.Bot/Requests/Available Methods/CopyMessageRequest.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        public int ReplyToMessageId { get; set; }
+
+        /// <inheritdoc />
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public IReplyMarkup ReplyMarkup { get; set; }
+
+        /// <summary>
+        /// Mode for parsing entities in the new caption. See formatting options for more details.
+        /// </summary>
+        public ParseMode ParseMode { get; set; }
```
Comment (Grepsy) [inline src/Telegram.Bot/Requests/Available Methods/CopyMessageRequest.cs:61]: Missing newline.

**Review by Fedorus — COMMENTED** [review]
This github interface... very unclear...

**Comment by Fedorus** [comment]
Squashed fixes, implemented proposed fixes.

**Comment by taha-ghadirian** [comment]
Hi, When does this PR apply?

File: src/Telegram.Bot/Requests/Available Methods/Admin Bot/UnpinChatMessageRequest.cs [hunk 2]
```diff
@@ -17,6 +17,12 @@ public class UnpinChatMessageRequest : RequestBase<bool>
         [JsonProperty(Required = Required.Always)]
         public ChatId ChatId { get; set; }
 
+        /// <summary>
+        /// Identifier of a message to unpin. If not specified, the most recent pinned message (by sending date) will be unpinned.
+        /// </summary>
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public int MessageId { get; set; }
```
Comment (taha-ghadirian) [inline src/Telegram.Bot/Requests/Available Methods/Admin Bot/UnpinChatMessageRequest.cs:24]: I Think we should pass the MessageId to the Constructor, Is that correct?

**Review by taha-ghadirian — CHANGES_REQUESTED** [review]
UnpinChatMessageRequest should be reviewed.

File: src/Telegram.Bot/Requests/Available Methods/Admin Bot/UnpinChatMessageRequest.cs (hunk 2 above)
Comment (karb0f0s) [reply src/Telegram.Bot/Requests/Available Methods/Admin Bot/UnpinChatMessageRequest.cs:24]: nope, according to [docs](https://core.telegram.org/bots/api#unpinchatmessage) `message_id ` is optional parameter. we should pass to constructor only required parameters.

**Comment by Fedorus** [comment]
Added baseUrl param to constructor (needed for private bot api servers) tested different methods with it... it works... 

EXCEPT DownloadFile method when server runs in --local mode. in that case it will give you local link to local file, for example on ubuntu: 
/home/ubuntu/telegram-bot-api/bin/<your bot token here>/documents/file_2
And we can`t modify DownloadFileAsync because .NetStandard 1.1 doesn`t support any file operations...

**Review by Muaath5 — CHANGES_REQUESTED** [review]
**Bot API 5.0 is working here perfectly**, And this pull request supports documentation.

This pull request support `CaptionEntities`,  `AllowSendingWithoutReply`, `Location`, `Pinned Messages`, Webhooks, Nearby chats, Dices, Anonymous Admins,**And all methods and types in Bot API 5.0**.

This pull request has Tests to Confirms that code is wroking successfully.

This pull request has all avaliable method and types, And it is very good.

**I suggest submiting this pull request.**

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 3]
```diff
@@ -412,17 +436,24 @@ public void StopReceiving()
             InputFileStream certificate = default,
             int maxConnections = default,
             IEnumerable<UpdateType> allowedUpdates = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            string ipAddress = default,
+            bool dropPendingUpdates = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:441]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 4]
```diff
@@ -412,17 +436,24 @@ public void StopReceiving()
...
+                DropPendingUpdates = dropPendingUpdates
             }, cancellationToken);
 
         /// <inheritdoc />
-        public Task DeleteWebhookAsync(CancellationToken cancellationToken = default)
-            => MakeRequestAsync(new DeleteWebhookRequest(), cancellationToken);
+        public Task DeleteWebhookAsync(
+            CancellationToken cancellationToken = default,
+            bool dropPendingUpdates = default
+        )
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:455]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 5]
```diff
@@ -445,14 +484,18 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             bool disableNotification = default,
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool allowSendingWithoutReply = default,
+            IEnumerable<MessageEntity> captionEntities = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:489]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 6]
```diff
@@ -478,14 +540,18 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
...
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:545]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 7]
```diff
@@ -502,18 +568,22 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            InputMedia thumb = default
+            InputMedia thumb = default,
+            bool allowSendingWithoutReply = default,
+            IEnumerable<MessageEntity> captionEntities = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:573]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 8]
```diff
@@ -527,16 +597,22 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            InputMedia thumb = default
+            InputMedia thumb = default,
+            bool allowSendingWithoutReply = default,
+            IEnumerable<MessageEntity> captionEntities = default,
+            bool disableContentTypeDetection = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:603]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 9]
```diff
@@ -546,12 +622,14 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             bool disableNotification = default,
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:626]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 10]
```diff
@@ -569,7 +647,9 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            InputMedia thumb = default
+            InputMedia thumb = default,
+            bool allowSendingWithoutReply = default,
+            IEnumerable<MessageEntity> captionEntities = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:652]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 11]
```diff
@@ -598,7 +680,9 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             bool disableNotification = default,
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool allowSendingWithoutReply = default,
+            IEnumerable<MessageEntity> captionEntities = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:685]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 12]
```diff
@@ -623,15 +709,19 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
...
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:714]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 13]
```diff
@@ -645,7 +735,8 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            InputMedia thumb = default
+            InputMedia thumb = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:739]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 14]
```diff
@@ -664,7 +756,8 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             IEnumerable<InputMediaBase> media,
             bool disableNotification = default,
             int replyToMessageId = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:760]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 15]
```diff
@@ -684,12 +778,14 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             ChatId chatId,
             bool disableNotification = default,
             int replyToMessageId = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:782]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 16]
```diff
@@ -701,14 +797,20 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             bool disableNotification = default,
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool allowSendingWithoutReply = default,
+            int heading = default,
+            int proximityAlertRadius = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:803]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 17]
```diff
@@ -723,15 +825,21 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            string foursquareType = default
+            string foursquareType = default,
+            string googlePlaceId = default,
+            string googlePlaceType = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:831]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 18]
```diff
@@ -744,14 +852,16 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            string vCard = default
+            string vCard = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:856]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 19]
```diff
@@ -772,12 +882,15 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             string explanation = default,
             ParseMode explanationParseMode = default,
             int? openPeriod = default,
-            DateTime? closeDate = default
+            DateTime? closeDate = default,
+            bool allowSendingWithoutReply = default,
+            IEnumerable<MessageEntity> explanationCaptionEntities = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:887]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 20]
```diff
@@ -797,12 +911,14 @@ public Task<User> GetMeAsync(CancellationToken cancellationToken = default)
             int replyToMessageId = default,
             IReplyMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            Emoji? emoji = default) =>
+            Emoji? emoji = default,
+            bool allowSendingWithoutReply = default) =>
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:915]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 21]
```diff
@@ -920,9 +1046,10 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
         public Task UnbanChatMemberAsync(
             ChatId chatId,
             int userId,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool onlyIfBanned = default
         ) =>
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1051]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 22]
```diff
@@ -997,10 +1124,12 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
             bool? canRestrictMembers = default,
             bool? canPinMessages = default,
             bool? canPromoteMembers = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool? isAnonymous = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1128]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 23]
```diff
@@ -1074,11 +1203,13 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
             ParseMode parseMode = default,
             bool disableWebPagePreview = default,
             InlineKeyboardMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            IEnumerable<MessageEntity> captionEntities = default
         ) =>
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1208]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 24]
```diff
@@ -1090,13 +1221,15 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
             ParseMode parseMode = default,
             bool disableWebPagePreview = default,
             InlineKeyboardMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            IEnumerable<MessageEntity> captionEntities = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1225]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 25]
```diff
@@ -1106,11 +1239,13 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
             string caption,
             InlineKeyboardMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
-            ParseMode parseMode = default
+            ParseMode parseMode = default,
+            IEnumerable<MessageEntity> captionEntities = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1243]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 26]
```diff
@@ -1120,11 +1255,13 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
...
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1259]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 27]
```diff
@@ -1181,11 +1318,17 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
             float latitude,
             float longitude,
             InlineKeyboardMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            float horizontalAccuracy = default,
+            int heading = default,
+            int proximityAlertRadius = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1324]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 28]
```diff
@@ -1194,11 +1337,17 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
...
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1343]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 29]
```diff
@@ -1274,7 +1423,8 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
             InlineKeyboardMarkup replyMarkup = default,
             CancellationToken cancellationToken = default,
             bool sendPhoneNumberToProvider = default,
-            bool sendEmailToProvider = default
+            bool sendEmailToProvider = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1427]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 30]
```diff
@@ -1347,12 +1498,14 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
             bool disableNotification = default,
             int replyToMessageId = default,
             InlineKeyboardMarkup replyMarkup = default,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            bool allowSendingWithoutReply = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1502]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 31]
```diff
@@ -1465,9 +1618,14 @@ await DownloadFileAsync(file.FilePath, destination, cancellationToken)
         /// <inheritdoc />
         public Task UnpinChatMessageAsync(
             ChatId chatId,
-            CancellationToken cancellationToken = default
+            CancellationToken cancellationToken = default,
+            int messageId = default
```
Comment (nesterenko-kv) [inline src/Telegram.Bot/TelegramBotClient.cs:1622]: The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

**Review by nesterenko-kv — COMMENTED** [review]
The default convention in .NET world is to setup `CancellationToken` at the end of the calling params.

File: src/Telegram.Bot/TelegramBotClient.cs (hunk 31 above)
Comment (Fedorus) [reply src/Telegram.Bot/TelegramBotClient.cs:1622]: Changing parameter order is considered breaking change, so it will happen in next major release

**Comment by akukharevich** [comment]
When will this PR be merged? :)

**Comment by Neurotoxin001** [comment]
Any ETA for new release?

**Review by karb0f0s — APPROVED** [review]

**Pull request merged**